newclient/client.go

```diff
@@ -0,0 +1,245 @@
+// Copyright © 2023 SECO Mind srl
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package newclient
+
+import (
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/astarte-platform/astarte-go/misc"
+)
+
+type Client struct {
+	baseURL            *url.URL
+	appEngineURL       *url.URL
+	housekeepingURL    *url.URL
+	pairingURL         *url.URL
+	realmManagementURL *url.URL
+	userAgent          string
+
+	httpClient *http.Client
+	token      string
+}
+
+type clientOption = func(c *Client) error
+
+// Finally, generics (actually, type constraints)
+type privateKeyProvider interface {
+	string | []byte
+}
+
+// The New function creates a new Astarte API client.
+// If no options are specified, the following is assumed:
+// - standard Astarte URL hierarchy
+// - standard HTTP client
+// - no JWT token (no call will be authorized)
+// - "astarte-go" as user agent
+// A production-ready client may be created using e.g.:
+// `client.New(client.WithBaseUrl("api.your-astarte.org"), client.WithToken("YOUR_JWT_TOKEN"))``
+func New(options ...clientOption) (*Client, error) {
+	// We start with a client with bare zero-valued fields
+	c := &Client{}
+
+	// Then we modify it according to user-provided options...
+	for _, f := range options {
+		err := f(c)
+		if err != nil {
+			return c, err
+		}
+	}
+
+	// ... and check if the result is valid
+	if err := validate(c); err != nil {
+		return c, err
+	}
+
+	// Finally, we add just a sprinkle of defaults and a new Client is born!
+	return setDefaults(c), nil
+}
+
+// The WithAppengineURL function allows to specify an
+// AppEngine URL different from the standard one (e.g. http://localhost:4000).
+// This is not recommendend in production.
+func WithAppengineURL(appEngineURL string) clientOption {
```

Comment:
The `AppEngine` is used throughout the rest of `astarte-go`, so the `E` should be capitalized here too